Feature: Rotating solid/sparse/top/bottom infills at any angles

pi-squared-studio opened the pull request:
A simple mod that allows you to expand the variations in the execution of internal infillings, as well as increase the strength of the tops and bottoms walls.

# Description

This is my second project for OrcaSlicer. I will be glad for any testing, comments, reviews and additions.
This mod expands the slicer's ability to generate internal fillings.

![image](https://github.com/user-attachments/assets/ebf0eef9-de16-4cf7-a710-83a63bbb6326)

The *Rotate __Solid__ Infill Direction* option is a replacement for the old version, where you could only choose to apply rotation or not. Now it is possible to rotate layers at different angles any infill pattern. This will be useful in the case of creating particularly strong walls of parts when it was previously impossible to achieve such a result.

The *Rotate __Sparse__ Infill Direction* option. Then I just freaked out and simultaneously implemented this feature. At first, I did not think what it would be useful for, but it is possible that it will find applications for the formation of laminar flows in air ducts, mufflers, or various filters. You should use this option with caution and always check the results in the preview. The final pattern will depend not only by the angle, but also on the size of the part. Always check and set this option to 0 if you are not using this function.

# Screenshots/Recordings/Graphs

Solid Infill with Monotonic pattern at 60 degrees:
![image](https://github.com/user-attachments/assets/8f729a8f-83a6-4c77-9c40-5744705fdcfb)
![image](https://github.com/user-attachments/assets/14789556-5c23-4c43-93f4-e0f8e2354a60)
![image](https://github.com/user-attachments/assets/95b6786d-c655-4065-9a70-7306d1304e98)

Solid Infill with Hilbert pattern at 30 degrees:
![image](https://github.com/user-attachments/assets/8ef3c501-bab2-4889-8f67-419591d22029)
![image](https://github.com/user-attachments/assets/0ea2d6d9-e817-4d97-ad0b-1f9a0c822c71)
![image](https://github.com/user-attachments/assets/917aa143-a631-4619-918a-82d71864ee7a)

Rotate Sparse Infill Examples:
![image](https://github.com/user-attachments/assets/de173161-2cea-4923-8985-2e559d1d16c5)
![image](https://github.com/user-attachments/assets/a940c97c-1991-4810-893c-5bccc1bbdf64)
![image](https://github.com/user-attachments/assets/9800ea29-722f-4d0c-8817-be694784345b)
![image](https://github.com/user-attachments/assets/95853e81-d4b1-4801-b46c-e2bfc524c7a3)
![image](https://github.com/user-attachments/assets/e388969e-2d2b-42ab-b758-4983230f05a7)
![image](https://github.com/user-attachments/assets/bc3a004e-da3d-4e54-986a-c074a68589fb)
![image](https://github.com/user-attachments/assets/a19ae4cc-e2ed-42ef-8404-9ec0cb2460f9)
![image](https://github.com/user-attachments/assets/667d3111-bdd6-4ed2-bd24-0ab0e120d710)
![image](https://github.com/user-attachments/assets/784c3cd9-a128-4495-9842-c99de6c73e0a)
etc

## Tests

Known problems:
- It is advisable to make an option to reset the value to 0 when any sparse infill selecting.
- Since the new parameters are now numeric, and it replaces the Boolean value, an incorrect value may be set when applying saved projects or presets. In most cases, it may be 1.

Ataraxia-Mechanica commented:
Works well, aside from infill combination rotates the infill by the amount of combined layer times instead of once. Is it possible to add a special case for this? Not familiar with Orca's codebase.
[Bed_Support_PET.zip](https://github.com/user-attachments/files/20718172/Bed_Support_PET.zip)

Noisyfox commented:
> Just disable the layer combination, or set the layer size to equal in the whole model.

I think what they want is automatically disable this when the rotation is enabled.

Noisyfox commented:
> > BambuStudio's zig zag infill.
> 
> It's a very interesting decision, but it was made in Bambu for nothing. This is a repeat of the Rectilinear preset, only with the addition of a parameter. I think you can just extend the effect of this preset to all the patterns. For the first time, this can only be implemented for Rectilinear and Lines... As for the names, it seems that sometimes they are not quite accurate, and do not reflect the essence of the function being performed, and sometimes a traditional and outdated name is given. If I am going to implement it, I will think about how best to name it, taking into account the past application. There may be a separate Zig-Zag preset.

Yeah I was confused af when they introduced the zig-zag & cross zag pattern.

pi-squared-studio commented:
A small announcement of upcoming changes. I've slightly tweaked the angle shaping algorithm. And if I don't get confused in the Github, then there will be another release soon.

![image](https://github.com/user-attachments/assets/1a7751a2-dfff-4456-9717-1fd43b89725f)

pi-squared-studio commented:
Please, check this additions: 
![image](https://github.com/user-attachments/assets/c268f9ef-0107-4b40-a66c-e19bfd8d9c11)

I rewrite some code.
Zig-Zag oscillations has added.  
Now you can easily specify the number of rotations of 1 cm infill lifting, or vice versa, specify how many centimeters 1 revolution (360 deg) of filling will be. The zero value indicate that infill will be rotate at each layer.

Ataraxia-Mechanica commented:
> I rewrite some code. Zig-Zag oscillations has added. Now you can easily specify the amount of rotations of 1 cm infill lifting, or vice versa, specify how many centimeters 1 revolution (360 deg) of filling will be. The zero value indicate that infill will be rotate at each layer.
> 
> That's works fine for me!

I think it would be better if it was in ```mm or %```, where % means the percentage of the layer height. Adding the same parameter for solid infill would be nice as well.

pi-squared-studio commented:
> I think it would be better if it was in `mm or %`, where % means the percentage of the layer height. Adding the same parameter for solid infill would be nice as well.

I've been thinking about it. It turns out to be very difficult to recalculate. Percentages need to be converted to millimeters or degrees, and 100 is not exactly a multiple of 90. And you can't just enter the degree sign from the keyboard.  Then you need to enter another parameter when you need to turn on rotation through each layer. But here you can do anything and very quickly.

pi-squared-studio commented:
Add:
- Top/Bottom surface rotation. I decided to extend the function to the rest of the infills. I remembered how it was necessary to orient the Monotonic pattern of the top surface to the vertical one, and for this I had to rotate the whole model by 45 degrees. The big surprise was that when the layer height changed, the infill changed its position. Now there is a strict binding to the parameter.
- Since there are a lot of infilling parameters accumulated, I decided to group them according to their purpose and order.

![image](https://github.com/user-attachments/assets/88b0bf1e-ecce-4b97-8b17-153c9e1d18e8)

Ataraxia-Mechanica commented:
> I've been thinking about it. It turns out to be very difficult to recalculate. Percentages need to be converted to millimeters or degrees, and 100 is not exactly a multiple of 90. And you can't just enter the degree sign from the keyboard. Then you need to enter another parameter when you need to turn on rotation through each layer. But here you can do anything and very quickly.

Here's an mock up of what I meant:
![image](https://github.com/user-attachments/assets/fae43ef2-b951-4ded-aff1-098c4f5f898a)

pi-squared-studio commented:
> Here's an mock up of what I meant:

Yes, it became a little clearer. But there are 2 points:
The height of the layer is too small to be attached to it. To specify, for example, 2 cm with a layer height of 0.2 mm, you will have to enter 10000%. This is extremely inconvenient. Then, when specifying this length, we abandon the binding to the layer number and attach to the height of the layer above the bed, which gives a smooth and predictable pattern that does not depend on the height of the layer. If we enter a percentage value, then it seems logical to take the total height of the model. Don't you think so?
It will be easy to make a clear turn or 2, 3... for the whole printed part, or only a certain turn for it.
Also, the dimensions in mm are very detailed and numerically large. It is more convenient to use centimeters if we are tied to height.

pi-squared-studio commented:
> Adding the same parameter for solid infill would be nice as well.

The solid fill is rotated only by a certain angle on each layer. I don't think it's worth tying it to height of model, since it doesn't contain a visible pattern, but only serves to create rigid shells. We can try to link the percentage value to the height of the layer, but what can it give us useful?

pi-squared-studio commented:
Another pack of usabilities.
+ Sector selection by %
+ Height of model selection by %
...

![image](https://github.com/user-attachments/assets/055fa001-95c4-4c7a-b343-aebd45eeb122)

afmenez commented on src/libslic3r/Fill/FillBase.hpp:
```diff
@@ -156,7 +156,7 @@ class Fill
         // Initial angle is undefined.
         angle(FLT_MAX),
-        rotate_angle(true),
+        rotate_angle(true), 
```
This file has only trailing spaces added, I believe it should be reversed.

pi-squared-studio replied:
OK. I know, I'll fix all this when I finish the code. Sometimes VisualStudio formats these lines, even if you don't make any changes.

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->label = L("Top infill direction");
     def->category = L("Strength");
-    def->tooltip = L("Rotate the solid infill direction by 90° for each layer.");
+    def->tooltip = L("Angle for top infill pattern, which controls the start or main direction of line" ".\n"
```
Why the `".\n"` is a separate string?

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloatOrPercent(45., false));
+
+    def = this->add("rotate_solid_infill_direction", coFloatOrPercent);
+    def->label = L("  Rotate solid infill");
```
Why the spaces at the beginning of the string?

afmenez replied:
Is this the correct way to indent? I don't see any other string indented like this... and I believe this alignment can break if the translated string is larger than the original.

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloatOrPercent(90., false));
+
+    def = this->add("rotate_sparse_infill_direction", coFloatOrPercent);
+    def->label = L("  Rotate sparse infill");
```
Why the spaces at the beginning of the string?

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloatOrPercent(0., false));
+
+    def = this->add("rotate_sparse_infill_height", coFloatOrPercent);
+    def->label = L("  ... at a height range of");
```
Why the spaces at the beginning of the string?

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloatOrPercent(0., false));
+
+    def = this->add("sparse_infill_zigzag_angle", coFloatOrPercent);
+    def->label = L("  Sparse infill zig-zag angle");
```
Why the spaces at the beginning of the string?

afmenez commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2321,32 +2321,121 @@ void PrintConfigDef::init_fff_params()
+    def->set_default_value(new ConfigOptionFloatOrPercent(0., false));
+
+    def = this->add("sparse_infill_zigzag_height", coFloatOrPercent);
+    def->label = L("  ... at a height range of");
```
Why the spaces at the beginning of the string?

pi-squared-studio replied:
![image](https://github.com/user-attachments/assets/39ee8c08-184b-4a03-bb34-77db64e6250d)
This is a piece of untouched code from another area. These gaps are met here.

afmenez reviewed (changes requested):
I have some issues with the label/tooltip strings...

pi-squared-studio commented:
A small announcement:
![image](https://github.com/user-attachments/assets/448ec889-7092-423d-9750-00daaf2bea14)
![image](https://github.com/user-attachments/assets/aac0cbdc-2780-45ed-b047-cf456c29401b)

after I correct the above remarks, I will post another update.

pi-squared-studio commented:
Ok. I have finished the work in this branch. Regarding the UI, I have provided my view on how these options should be called. But I don't mind if they have a different name and are presented in other menus. I think that the author should resolve this point further.

yw4z commented:
looks like twist effect on some cad apps if value lower than 5. this would look cool when printing translucent things like lighting products

pi-squared-studio commented:
I don't know what the reason is. I fixed all the errors and rewrote the code so that it could not affect the original, or even hypothetically modified, main repository in any way. The branch when strings only added not changed, but the branch still causes a conflict. On my side, the code is compiled. Perhaps this is a repository conflict, the one that is available to me and the one that is the latest unavailable version.

pi-squared-studio commented:
There were some minor problems with the code. I will redo the project soon.

discip commented:
What exactly do you mean?

<img width="356" height="94" alt="image" src="https://github.com/user-attachments/assets/0dcf47d9-2e20-4754-83e0-94dc426206d3" />

<br>
<br>

As noted here:

<img width="398" height="122" alt="image" src="https://github.com/user-attachments/assets/d3b18637-43ff-44a2-9780-dd1978887a83" />

It can be calculated automatically, but that doesn't always work. 😊

discip commented:
Great!
Which PR is it?

ps:
Have you by any chance considered the direction of the **`support interface`** as well?
Today, I stumbled upon exactly that.

pi-squared-studio commented:
@discip Here is the PR https://github.com/SoftFever/OrcaSlicer/pull/10314 and it has very huge list of changes
